tensorflow/core/kernels/mkl_qmatmul_op.cc

```diff
@@ -0,0 +1,875 @@
+      float max_output_value;
+      MklQuantizationRangeForMultiplication<quint8, qint8, qint32>(
+          min_input, max_input, min_weight, max_weight, &min_output_value,
+          &max_output_value);
+      float scale_int32 =
+          std::max(std::abs(min_output_value), std::abs(max_output_value));
+      float scale_eightbit =
+          std::max(std::abs(min_freezed_output), std::abs(max_freezed_output));
+      float scale = 1.0;
+      if (std::is_same<Toutput, quint8>::value)
+        scale = scale_int32 / scale_eightbit / static_cast<float>(1 << 23);
+      else
+        scale = scale_int32 / scale_eightbit / static_cast<float>(1 << 24);
+
+      std::vector<float> output_scale;
+      output_scale.push_back(scale);
+      params.post_op_params.push_back({"output_scale", output_scale});
+    }
+  }
+
+  // This function handles bias conversion and compensation
+  // for MIN_FIRST and SCALE mode
+  // If input is quantized via MIN_FIRST
+  //  Bs32=QaQwB𝑓32 + QaMin(A𝑓32) Ws8*1
+  // If input is quantized via SCALE
+  //  Bs32=QaQwB𝑓32
+  //    where QaQw is the multiply of Qa and Qw,
+  //      also is called output quantize scale
+  Tbias* GetBiasHandle(
+      OpKernelContext* context,
+      std::shared_ptr<mkldnn::inner_product_forward::primitive_desc>&
+          mkldnn_matmul_fwd_pd,
+      const Tensor& bias_tensor, const Tensor& weight_tensor) {
+    // If the bias is int32, it means the bias is already be converted offline.
+    // and it can be added to matmul output directly.
+    if (std::is_same<Tbias, qint32>::value) {
+      return static_cast<Tbias*>(
+          const_cast<Tbias*>(bias_tensor.flat<Tbias>().data()));
+    } else {
+      // If the bias is fp32, then need to calculate the bias
+      const float min_input = context->input(3).flat<float>()(0);
+      const float max_input = context->input(4).flat<float>()(0);
+      const float min_weight = context->input(5).flat<float>()(0);
+      const float max_weight = context->input(6).flat<float>()(0);
+
+      std::vector<mkldnn::primitive> net;
+      float out_scale;
+      // If the bias is float and input quantize is MIN_FIRST
+      // bias has to be compensated with
+      // Bs32=QaQwB𝑓32 + QaMin(A𝑓32) Ws8*1
+      if (mode_ == QUANTIZE_MODE_MIN_FIRST) {
+        int k = weight_tensor.dim_size(0);
+        int n = weight_tensor.dim_size(1);
+        float* comp_bias = GetCompBiasBuffer(n);
+
+        qint8* wt_buf = static_cast<qint8*>(
+            const_cast<qint8*>(weight_tensor.flat<qint8>().data()));
+
+        const float* bias_buf = static_cast<float*>(
+            const_cast<float*>(bias_tensor.flat<float>().data()));
+
+        float qa_amin = 255 * min_input / (max_input - min_input);
+
+        out_scale = (255.0 * 127.0) /
+                    ((max_input - min_input) *
+                     std::max(std::abs(max_weight), std::abs(min_weight)));
+
+#pragma omp parallel for schedule(static)
+        for (int j = 0; j < n; j++) {
```

Comment:
```suggestion
        for (int j = 0; j < n; ++j) {
```
Same with the `i` loop.